apps/prairielearn/src/lib/editors.ts

```diff
@@ -1002,6 +1021,53 @@ export class QuestionAddEditor extends Editor {
   }
 }
 
+export class QuestionModifyEditor extends Editor {
+  private question: Question;
+  private origHash: string;
+  private files: Record<string, string>;
+
+  constructor(
+    params: BaseEditorOptions & {
+      files: Record<string, string>;
+    },
+  ) {
+    super(params);
+
+    this.question = params.locals.question;
+    this.files = params.files;
+    this.description = `Modify question ${this.question.qid}`;
+  }
+
+  async write() {
+    assert(this.question.qid, 'question.qid is required');
+
+    const questionPath = path.join(this.course.path, 'questions', this.question.qid);
+
+    // Validate that all file paths don't escape the question directory.
+    for (const filePath of Object.keys(this.files)) {
+      if (!contains(questionPath, filePath)) {
+        throw new Error(`Invalid file path: ${filePath}`);
+      }
+    }
+
+    // Note that we deliberately only modify files that were provided. We don't
+    // try to delete "excess" files that aren't in the `files` object because the
+    // user might have added extra files of their own, e.g. in the `tests` or
+    // `clientFilesQuestion` directory. We don't want to remove them, and we also
+    // don't want to mandate that the caller must always read all existing files
+    // and provide them in the `files` object.
```

Comment:
Would it be helpful to also allow a `filesToBeDeleted` parameter (with a better name) for this function? I can imagine that we might sometimes want to delete `server.py` during AI generation if the LLM decides that `server.py` is no longer needed for this question.

Reply:
Hah, interesting thought. Tagging in @jlevine18 for the broader question of deletions. There's a specific scenario I'm worrying about:

- Question is generated by AI.
- Human manually adds/edits `server.py`.
- Human uses AI to further revise the question.
- AI (for whatever reason) erroneously decides that `server.py` is no longer needed and deletes it.

Maybe the ability to undo/reset to a particular past revision is sufficient to resolve this?

Reply:
I'm going to avoid adding this for now. We can revisit this later, if needed.